Handle NaNs in peaks

jonathanjfshaw opened the pull request:
Fixes #11

jonathanjfshaw committed:
Handle NaNs in peaks

jonathanjfshaw committed:
Include RollingDetect too

jonathanjfshaw committed:
Remove extra blank line

bertrandlalo commented on timeflux_dsp/nodes/peaks.py:
```diff
@@ -128,6 +128,11 @@ def update(self):
             self.i.data = self.i.data.take([0], axis=1)
 
+        # If there are no no-NaN rows in the column we're considering, there is nothing to do.
+        self.i.data = self.i.data.dropna()
```
I'd prefer the user to drop NaNs outside this Node. 
We could add a warning when receiving a chunk with NaNs and skip it if it makes the update failing.

bertrandlalo reviewed (changes requested):
If the data has NaN, the user should add a 'Dropna' Node just before. I don't like it when a node does more than what it should, ie. here the node's purpose is to detect peaks, not to drop NaNs. 
Again, we could choose to add a `dropna` parameter but this could be the case in any node, and in my opinion, the user should handle its NaN separately.

jonathanjfshaw commented:
NaNs don't cause the update loop to fail, just a RunTimeWarning to be thrown.

https://github.com/timeflux/timeflux/pull/59 provides a dropna() node.

> . I don't like it when a node does more than what it should, ie. here the node's purpose is to detect peaks, not to drop NaNs.

It seems to me that NaNs are a reasonable input to a peak detection algorithm, and we should handle them in a reasonable way.

I see 4 options: 
1) keep the current behavior of allowing NaNs and ignoring them, but change the code so that a RunTimeWarning is not generated.
2) treat a NaN as a possible peak/valley, meaning we no longer know what the value of the last peak/valley was and have to resume looking for peaks/valleys as if starting from the beginning.
3) provide a parameter to choose between option 1 and option 2
4) Document that we don't allow NaNs here, e.g. that the user should use a Missing node from https://github.com/timeflux/timeflux/pull/59, and raise an Exception if a NaN is supplied.

At the moment we are sort of half-way between option 1 and option 4: NaNs actually work fine, but we get an irrelevant runtime warning.

mesca commented:
I agree with @bertrandlalo, NaNs should be filtered in a dedicated node. 

At the very beginning, we should emit a `logging.warn` (not a  RunTimeWarning) message if the dataframe contains NaNs and skip the whole `update`.

In the current situation, even if it "works", the behavior is undefined. At the same time, I would hate to have a general exception in a middle of an acquisition because of one NaN :)

jonathanjfshaw committed:
Skip chunk for NaN

jonathanjfshaw commented:
I implemented @mesca wishes in #12. But I think this approach makes no sense.

> we should emit a logging.warn (not a RunTimeWarning) message if the dataframe contains NaNs and skip the whole update.
> ...
> I would hate to have a general exception in a middle of an acquisition because of one NaN :)

 With this approach which I have now implemented in #12 we are carrying on peak detection despite NaNs, but throwing away values that happen to fall in the same chunk as the NaN. But from a peak detection perspective, theres nothing special about whether a value falls in the same chunk or not.

If we wish to honor the idea of NaNs as "undefined" values which should really be filtered out before peak detection, then we should do my option (2) above:

> 2. treat a NaN as a possible peak/valley, meaning we no longer know what the value of the last peak/valley was and have to resume looking for peaks/valleys as if starting from the beginning.

I have implemented this in #18. This seems to have the characteristics you desire:
- it handles NaNs in a defined and principled fashion (as an UNKNOWN value), rather than simply dropping them
- it won't crash a whole acquisition session
